autoscheduler/scraper/banner_requests.py

```diff
@@ -149,24 +148,58 @@ def get_departments(self, amount: int = 300) -> List[Dict]:
 
         return depts
 
-    async def search(self, depts: List[str], amount: int = 750) -> List[List[Dict]]:
+    async def search(self, depts_terms: List[Tuple[str, str]],
+                     sem: asyncio.Semaphore,
+                     parse_all_courses: Callable[[list], list],
+                     amount: int = 750
+                    ) -> List[List[Dict]]:
         """ Concurrently retrieves all of the given departments and returns them as
             a list of course-lists, with each index corresponding to the courses/sections
             for a department.
         """
 
         loop = asyncio.get_running_loop()
 
-        async def perform_search(dept: str):
-            result = None
+        courses_set = set()
+        instructors_set = set()
+
+        async def perform_search(dept: str, term: str):
+            result = []
+
             async with ClientSession(loop=loop) as session:
-                session_id = await self.create_session(session)
+                course_list = []
+
+                retry_max = 10
+                for i in range(1, retry_max + 1):
+                    try:
+                        # Must limit the amoutn of concurrent requests, otherwise will get
+                        # a "too many file descriptors in select()" error from aiohttp
+                        async with sem:
+                            print(f"Starting {dept} {term}")
+                            session_id = await self.create_session(session, term)
+
+                            course_list = await self.get_courses(session, session_id,
+                                                                 dept, term, amount)
+
+                        # We only want to limit the requests, not parsing, so call this
+                        # outside of the semaphore
+                        ret = parse_all_courses(course_list, term, courses_set,
+                                                instructors_set)
+
+                        if ret is False:
+                            continue # Failure (course_list was none), retry

```

Comment:
Why would you want to retry here? You seem to be catching all the network errors in the `except` block, and if the request was valid, shouldn't continuing here just return the same results?

Reply:
It's a really weird edge case that only occasionally happens, but the request will return without throwing an error but it returns `None` when it shouldn't, in which case we need to retry. That being said we should probably handle this case in `perform_search` rather than `parse_all_courses`.

Reply:
To be honest, we should just change this line to `if course_list is None: continue`, then remove the `Union[bool, list]` or whatever it is from `parse_all_courses`, as there's no reason we shouldn't be checking for this case in `perform_search`.